diff -r -c -C 3 libsrc.orig/farz88/Makefile libsrc/farz88/Makefile
*** libsrc.orig/farz88/Makefile	Mon Dec  9 19:02:04 2002
--- libsrc/farz88/Makefile	Mon Dec  9 19:02:35 2002
***************
*** 6,12 ****
  
  all:
  	cd strings; $(MAKE)
! 	z80asm -d -ns -nm -Mo -x../farz88 @farz88.lst
  
  clean:
  	$(RM) *.o *~
--- 6,12 ----
  
  all:
  	cd strings; $(MAKE)
! 	z80asm -d -ns -nm -Mo -x/farz88 @farz88.lst
  
  clean:
  	$(RM) *.o *~
Only in libsrc/farz88: Makefile.orig
diff -r -c -C 3 libsrc.orig/fcntl/dummy/Makefile libsrc/fcntl/dummy/Makefile
*** libsrc.orig/fcntl/dummy/Makefile	Mon Dec  9 19:02:04 2002
--- libsrc/fcntl/dummy/Makefile	Mon Dec  9 19:02:36 2002
***************
*** 4,10 ****
  	@echo ''
  	@echo '---> Building NDOS library <---'
  	@echo ''
! 	z80asm -d -ns -nm -Mo -x../../ndos @ndoslist
  
  clean:
  	$(RM) *.sym *.map zcc_opt.def *.o*
--- 4,10 ----
  	@echo ''
  	@echo '---> Building NDOS library <---'
  	@echo ''
! 	z80asm -d -ns -nm -Mo -x//ndos @ndoslist
  
  clean:
  	$(RM) *.sym *.map zcc_opt.def *.o*
Only in libsrc/fcntl/dummy: Makefile.orig
diff -r -c -C 3 libsrc.orig/genmath/Makefile libsrc/genmath/Makefile
*** libsrc.orig/genmath/Makefile	Mon Dec  9 19:02:04 2002
--- libsrc/genmath/Makefile	Mon Dec  9 19:02:36 2002
***************
*** 20,26 ****
  all: genmath
  
  genmath: $(OBJECTS)
! 	z80asm -d -ns -nm -Mo -x../gen_math @genlist
  
  .c.o:
  	zcc -vn -make-lib $*.c
--- 20,26 ----
  all: genmath
  
  genmath: $(OBJECTS)
! 	z80asm -d -ns -nm -Mo -x/gen_math @genlist
  
  .c.o:
  	zcc -vn -make-lib $*.c
Only in libsrc/genmath: Makefile.orig
diff -r -c -C 3 libsrc.orig/malloc/Makefile libsrc/malloc/Makefile
*** libsrc.orig/malloc/Makefile	Mon Dec  9 19:02:04 2002
--- libsrc/malloc/Makefile	Mon Dec  9 19:02:36 2002
***************
*** 9,15 ****
  
  malloc: calloc.o realloc_down.o
  	cd strings; $(MAKE)
! 	z80asm -d -ns -nm -Mo -x../malloc @malllist
  
  .c.o:
  	zcc -vn -make-lib $*.c
--- 9,15 ----
  
  malloc: calloc.o realloc_down.o
  	cd strings; $(MAKE)
! 	z80asm -d -ns -nm -Mo -x/malloc @malllist
  
  .c.o:
  	zcc -vn -make-lib $*.c
Only in libsrc/malloc: Makefile.orig
diff -r -c -C 3 libsrc.orig/net/Makefile libsrc/net/Makefile
*** libsrc.orig/net/Makefile	Mon Dec  9 19:02:04 2002
--- libsrc/net/Makefile	Mon Dec  9 19:02:37 2002
***************
*** 23,30 ****
  all: net
  
  net: $(OBJECTS)
! 	z80asm -d -ns -nm -Mo -x../net @zsock_list
! 	z80asm -d -ns -nm -Mo -x../netdev @zsockdev_list
  
  .c.o:
  	zcc -vn -Wn43 -make-lib $*.c 
--- 23,30 ----
  all: net
  
  net: $(OBJECTS)
! 	z80asm -d -ns -nm -Mo -x/net @zsock_list
! 	z80asm -d -ns -nm -Mo -x/netdev @zsockdev_list
  
  .c.o:
  	zcc -vn -Wn43 -make-lib $*.c 
Only in libsrc/net: Makefile.orig
diff -r -c -C 3 libsrc.orig/z80_crt0s/Makefile libsrc/z80_crt0s/Makefile
*** libsrc.orig/z80_crt0s/Makefile	Mon Dec  9 19:02:04 2002
--- libsrc/z80_crt0s/Makefile	Mon Dec  9 19:02:37 2002
***************
*** 8,18 ****
  
  crt0s:
  	@$(RM) crt0_long/l_long_case.o
! 	z80asm -d -ns -nm -Mo -x../z80_crt0 @crt0list
  
  crtz80s183:
  	@$(RM) crt0_long/l_long_case.o
! 	z80asm -d -DZ80S183 -ns -nm -Mo -x../z80s183_crt0 @crt0list
  
  
  clean:
--- 8,18 ----
  
  crt0s:
  	@$(RM) crt0_long/l_long_case.o
! 	z80asm -d -ns -nm -Mo -x/z80_crt0 @crt0list
  
  crtz80s183:
  	@$(RM) crt0_long/l_long_case.o
! 	z80asm -d -DZ80S183 -ns -nm -Mo -x/z80s183_crt0 @crt0list
  
  
  clean:
Only in libsrc/z80_crt0s: Makefile.orig
diff -r -c -C 3 libsrc.orig/z88/Makefile libsrc/z88/Makefile
*** libsrc.orig/z88/Makefile	Mon Dec  9 19:02:04 2002
--- libsrc/z88/Makefile	Mon Dec  9 19:02:37 2002
***************
*** 29,35 ****
  	cd fdstdio ; $(MAKE)
  
  z88: $(OBJECTS)
! 	z80asm -d -ns -nm -Mo -x../z88 @z88list
  
  .c.o:
  	zcc -vn -make-lib $*.c
--- 29,35 ----
  	cd fdstdio ; $(MAKE)
  
  z88: $(OBJECTS)
! 	z80asm -d -ns -nm -Mo -x/z88 @z88list
  
  .c.o:
  	zcc -vn -make-lib $*.c
Only in libsrc/z88: Makefile.orig
diff -r -c -C 3 libsrc.orig/z88math/Makefile libsrc/z88math/Makefile
*** libsrc.orig/z88math/Makefile	Mon Dec  9 19:02:04 2002
--- libsrc/z88math/Makefile	Mon Dec  9 19:02:38 2002
***************
*** 7,13 ****
  
  
  z88math:
! 	z80asm -d -ns -nm -Mo -x../z88_math @z88list
  
  
  clean:
--- 7,13 ----
  
  
  z88math:
! 	z80asm -d -ns -nm -Mo -x/z88_math @z88list
  
  
  clean:
Only in libsrc/z88math: Makefile.orig
