--- lwtools-4.22/lwasm/insn_rel.c.bak	Sat Dec 23 04:05:00 2023
+++ lwtools-4.22/lwasm/insn_rel.c	Mon Feb 26 00:32:32 2024
@@ -84,9 +84,9 @@
 		{
 			if (strncasecmp(*p, "?RTS", 4) == 0)
 			{
+				line_t *cl = l;
 				(*p) += 4;
 
-				line_t *cl = l;
 				for (cl = cl->prev; cl; cl = cl->prev)
 				{
 					if (cl->insn == -1)
--- lwtools-4.22/lwasm/list.c.bak	Sat Dec 23 04:05:00 2023
+++ lwtools-4.22/lwasm/list.c	Mon Feb 26 01:05:44 2024
@@ -84,8 +84,8 @@
 		}
 		if (cl -> noexpand_start)
 		{
-			obytelen = 0;
 			int nc = 0;
+			obytelen = 0;
 			for (nl = cl; nl; nl = nl -> next)
 			{
 				if (nl -> noexpand_start)
@@ -232,6 +232,7 @@
 		}
 
 		/* display cycle counts */
+		{
 		char s[64] = "";
 		if (CURPRAGMA(cl, PRAGMA_C) || CURPRAGMA(cl, PRAGMA_CD))
 		{
@@ -258,6 +259,7 @@
 		}
 
 		if (of) fprintf(of, "%-8s", s);
+		}
 
 		if (CURPRAGMA(cl, PRAGMA_CT)) 
 		{
--- lwtools-4.22/lwasm/lwasm.c.bak	Sat Dec 23 04:05:00 2023
+++ lwtools-4.22/lwasm/lwasm.c	Mon Feb 26 01:12:49 2024
@@ -306,22 +306,24 @@
 /* parse unit test input data from comment field */
 void lwasm_parse_testmode_comment(line_t *l, lwasm_testflags_t *flags, lwasm_errorcode_t *err, int *len, char **buf)
 {
+	char* s = strstr(l -> ltext, ";.");
+	char* t = strstr(s, ":");
 	*flags = 0;
 
 	if (!l)
 		return;
 
-	char* s = strstr(l -> ltext, ";.");
+	s = strstr(l -> ltext, ";.");
 	if (s == NULL) return;
 
-	char* t = strstr(s, ":");
+	t = strstr(s, ":");
 	if (t == NULL)
 	{
 		/* parse: ;.8E0FCE (emitted code) */
+		int i;
 
 		if (buf == NULL) return;
 
-		int i;
 		*flags = TF_EMIT;
 
 		s = s + 2;	/* skip ;. prefix */
@@ -350,9 +352,9 @@
 	else
 	{
 		/* parse: ;.E:1000 or ;.E:7 (warnings or errors) */
+		char ch = toupper(*(t - 1));
 		*flags = TF_ERROR;
 
-		char ch = toupper(*(t - 1));
 		if (ch != 'E') lwasm_error_testmode(l, "bad test data (expected E: flag)", 1);
 		sscanf(t + 1, "%d", (int*) err);
 	}
@@ -372,12 +374,12 @@
 		lwasm_parse_testmode_comment(l, &flags, &testmode_error_code, NULL, NULL);
 		if (flags == TF_ERROR)
 		{
+			char buf[128];
 			l -> len = 0;	/* null out bogus line */
 			l -> insn = -1;
 			l -> err_testmode = error_code;
 			if (testmode_error_code == error_code) return;		/* expected error: ignore and keep assembling */
 
-			char buf[128];
 			sprintf(buf, "wrong error code (%d)", error_code);
 			lwasm_error_testmode(l, buf, 0);
 			return;
@@ -414,11 +416,10 @@
 {
 	char errbuff[1024];
 	char f[128];
+	va_list args;
 
 	sprintf(f, "%s %s", lwasm_lookup_error(err), fmt);
 
-	va_list args;
-
 	va_start(args, fmt);
 
 	(void) vsnprintf(errbuff, 1024, f, args);
@@ -1237,11 +1238,12 @@
 	
 	for (cl = as -> line_head; cl; cl = cl -> next)
 	{
+		char* s;
 		if (!(cl -> err) && !(cl -> warn))
 			continue;
 
 		// trim "include:" if it appears
-		char* s = cl->linespec;
+		s = cl->linespec;
 		if ((strlen(s) > 8) && (s[7] == ':')) s += 8;
 		while (*s == ' ') s++;
 
--- lwtools-4.22/lwasm/pseudo.c.bak	Sat Dec 23 04:05:00 2023
+++ lwtools-4.22/lwasm/pseudo.c	Mon Feb 26 01:13:58 2024
@@ -979,11 +979,12 @@
 
 PARSEFUNC(pseudo_parse_reorg)
 {
+	line_t *cl;
 	lw_expr_t e = NULL;
 
 	l -> len = 0;
 
-	line_t *cl = l;
+	cl = l;
 	for (cl = cl -> prev; cl; cl = cl -> prev)
 	{
 		if (cl -> insn == -1) continue;
--- lwtools-4.22/lwcc/cpp-main.c.bak	Sat Dec 23 04:05:00 2023
+++ lwtools-4.22/lwcc/cpp-main.c	Mon Feb 26 01:15:22 2024
@@ -112,8 +112,8 @@
 
 int main(int argc, char **argv)
 {
-	program_name = argv[0];
 	int retval = 0;
+	program_name = argv[0];
 	
 	input_files = lw_stringlist_create();
 	includedirs = lw_stringlist_create();
--- lwtools-4.22/lwcc/preproc.c.bak	Sat Dec 23 04:05:00 2023
+++ lwtools-4.22/lwcc/preproc.c	Mon Feb 26 01:15:56 2024
@@ -1064,11 +1064,11 @@
 {
 	int c;
 	int c2;
+	int rv = 0;
 	
 	if (**t == 0)
 		return 0;
 	c = *(*t)++;
-	int rv = 0;
 	
 	switch (c)
 	{
