--- tibasic.c.bak	Wed Aug 12 03:44:53 2009
+++ tibasic.c	Sat Mar 27 23:59:34 2010
@@ -66,7 +66,7 @@
 
 #if defined(__GNUC__) && defined(__APPLE_CC__)
 #define __ATTRIBUTE_WEAK__
-#elif defined(__GNUC__)
+#elif defined(__GNUC__) && !defined(__MORPHOS__)
 #define __ATTRIBUTE_WEAK__ __attribute__((weak))
 #else
 #define __ATTRIBUTE_WEAK__
--- bin2c.pl.bak	Fri Aug  7 04:42:01 2009
+++ bin2c.pl	Sun Mar 28 00:16:25 2010
@@ -22,7 +22,7 @@
 
 open(OUT, '>', $out_filename) or die "can't open $out_filename: $!";
 print OUT "/* automatically generated from $in_filename */\n";
-print OUT "const unsigned char ${out_name}[] = {\n";
+print OUT "const unsigned char ${out_name} [] = {\n";
 my $i;
 for ($i = 0; $i < @in; $i++)
 {
