*** load_med.cpp.old	2017-12-09 18:21:56.173217851 +0200
--- load_med.cpp	2017-12-09 18:27:37.801021787 +0200
***************
*** 487,497 ****
  
  	if ((!lpStream) || (dwMemLength < 0x200)) return FALSE;
  	pmmh = (MEDMODULEHEADER *)lpStream;
! 	if (((pmmh->id & 0x00FFFFFF) != 0x444D4D) || (!pmmh->song)) return FALSE;
  	// Check for 'MMDx'
  	DWORD dwSong = bswapBE32(pmmh->song);
  	if ((dwSong >= dwMemLength) || (dwSong + sizeof(MMD0SONGHEADER) >= dwMemLength)) return FALSE;
! 	version = (signed char)((pmmh->id >> 24) & 0xFF);
  	if ((version < '0') || (version > '3')) return FALSE;
  #ifdef MED_LOG
  	Log("\nLoading MMD%c module (flags=0x%02X)...\n", version, bswapBE32(pmmh->mmdflags));
--- 487,497 ----
  
  	if ((!lpStream) || (dwMemLength < 0x200)) return FALSE;
  	pmmh = (MEDMODULEHEADER *)lpStream;
! 	if (((bswapBE32(pmmh->id) & 0xFFFFFF00) != 0x4D4D4400) || (!pmmh->song)) return FALSE;
  	// Check for 'MMDx'
  	DWORD dwSong = bswapBE32(pmmh->song);
  	if ((dwSong >= dwMemLength) || (dwSong + sizeof(MMD0SONGHEADER) >= dwMemLength)) return FALSE;
! 	version = (signed char)(bswapBE32(pmmh->id) & 0xFF);
  	if ((version < '0') || (version > '3')) return FALSE;
  #ifdef MED_LOG
  	Log("\nLoading MMD%c module (flags=0x%02X)...\n", version, bswapBE32(pmmh->mmdflags));
***************
*** 631,637 ****
  	// Reading play sequence
  	if (version < '2')
  	{
! 		UINT nbo = pmsh->songlen >> 8;
  		if (nbo >= MAX_ORDERS) nbo = MAX_ORDERS-1;
  		if (!nbo) nbo = 1;
  		memcpy(Order, pmsh->playseq, nbo);
--- 631,637 ----
  	// Reading play sequence
  	if (version < '2')
  	{
! 		UINT nbo = bswapBE16(pmsh->songlen);
  		if (nbo >= MAX_ORDERS) nbo = MAX_ORDERS-1;
  		if (!nbo) nbo = 1;
  		memcpy(Order, pmsh->playseq, nbo);
***************
*** 676,682 ****
  				{
  					for (UINT i=0; i<n; i++)
  					{
! 						UINT seqval = pmps->seq[i] >> 8;
  						if ((seqval < wNumBlocks) && (nOrders < MAX_ORDERS-1))
  						{
  							Order[nOrders++] = seqval;
--- 676,682 ----
  				{
  					for (UINT i=0; i<n; i++)
  					{
! 						UINT seqval = bswapBE16(pmps->seq[i]);
  						if ((seqval < wNumBlocks) && (nOrders < MAX_ORDERS-1))
  						{
  							Order[nOrders++] = seqval;
***************
*** 859,866 ****
  		#endif
  			const MMD1BLOCKINFO *pbi = NULL;
  			BYTE *pcmdext = NULL;
! 			lines = (pmb->lines >> 8) + 1;
! 			tracks = pmb->numtracks >> 8;
  			if (!tracks) tracks = m_nChannels;
  			if ((Patterns[iBlk] = AllocatePattern(lines, m_nChannels)) == NULL) continue;
  			PatternSize[iBlk] = (WORD)lines;
--- 859,866 ----
  		#endif
  			const MMD1BLOCKINFO *pbi = NULL;
  			BYTE *pcmdext = NULL;
! 			lines = bswapBE16(pmb->lines) + 1;
! 			tracks = bswapBE16(pmb->numtracks);
  			if (!tracks) tracks = m_nChannels;
  			if ((Patterns[iBlk] = AllocatePattern(lines, m_nChannels)) == NULL) continue;
  			PatternSize[iBlk] = (WORD)lines;
