*** include/sys/types.h.orig	Wed Nov 03 10:12:08 1993
--- include/sys/types.h	Mon Aug 10 15:29:16 1992
***************
*** 42,49 ****
  typedef	unsigned long	u_long;
  typedef	unsigned short	ushort;		/* Sys V compatibility */
  
- typedef long		BPTR;
- 
  typedef	char *	caddr_t;		/* core address */
  typedef	long	daddr_t;		/* disk address */
  typedef	short	dev_t;			/* device number */
--- 42,47 ----
