*** dice:libsrc/unix/lstat.c	Sun Nov 22 01:29:52 1992
--- lstat.c	Fri May 05 06:23:29 1995
***************
*** 34,43 ****
--- 34,44 ----
      if (readlink(name, buf, sizeof(buf)) >= 0) {
          clrmem(st, sizeof(*st));
  
          st->st_mode = S_IFLNK | S_IREAD | S_IWRITE;
          st->st_ctime = st->st_mtime = time(NULL);
+         st->st_dev = (long)DeviceProc(UnixToAmigaPath(name));
          return(0);
      } else {
          return(stat(UnixToAmigaPath(name), st));
      }
  }
