*** timidity/config.h.old	2012-01-16 00:01:05.000000000 +0200
--- timidity/config.h	2013-03-12 18:26:29.175071591 +0200
***************
*** 160,168 ****
--- 160,174 ----
  
  /* You could specify a complete path, e.g. "/etc/timidity.cfg", and
     then specify the library directory in the configuration file. */
+ #ifdef __amigaos4__
+ #define CONFIG_FILE	"PROGDIR:timidity.cfg"
+ #define CONFIG_FILE_ETC "ENV:timidity.cfg"
+ #else
  #define CONFIG_FILE	"timidity.cfg"
  #define CONFIG_FILE_ETC "/etc/timidity.cfg"
+ #endif
  
+ #ifndef __amigaos4__
  #if defined(__WIN32__) || defined(__OS2__)
  #define DEFAULT_PATH	"C:\\TIMIDITY"
  #else
***************
*** 171,176 ****
--- 177,183 ----
  #define DEFAULT_PATH2	"/usr/local/share/timidity"
  #define DEFAULT_PATH3	"/usr/local/lib/timidity"
  #endif
+ #endif
  
  /* These affect general volume */
  #define GUARD_BITS 3
