--- /home/guest/tox/tox.c	Thu Aug 21 00:17:20 2003
+++ tox.c	Thu Aug 21 01:08:10 2003
@@ -489,6 +489,7 @@
 	return r;
 }
 
+#ifdef HTML
 void find_string (parser_state *parser)
 {
 	char_t c;
@@ -507,6 +508,7 @@
         parser->find_string_current = 0;
     }
 }
+#endif
 
 void skip_dtd (parser_state *parser)
 {
@@ -656,6 +658,7 @@
                             parser->angleBracketLevel = 0;
                         }
                         break;
+#ifdef HTML
                     case SPECIAL_FIND_STRING:
                         find_string (parser);
                         transition = 1;
@@ -663,6 +666,7 @@
                             parser->state = state_table[parser->state][i].next_state;
                         }
                         break;
+#endif
 
                 }
                 break;
