--- t:wierd/wierd.c	Fri Jul 20 23:48:10 2001
+++ wierd.c	Tue Feb 26 00:08:57 2002
@@ -97,13 +97,22 @@
 
  if (argc < 2)		/* Not enough arguments--don't know what to do */
 	{
-	 fprintf (stderr, "Wierd interpreter, Atari version\n\nError - no file specified:\n\t%s <Wierd Filename>\nFehler - keine Datei angegeben:\n\t%s <Wierd Dateiname>\n", argv[0]);
+	#ifdef _AMIGA
+	 fprintf (stderr, "Wierd interpreter, Amiga version\n\nError - no file specified:\n\t%s <Wierd Filename>\nFehler - keine Datei angegeben:\n\t%s <Wierd Dateiname>\n",
+	#else
+	 fprintf (stderr, "Wierd interpreter, Atari version\n\nError - no file specified:\n\t%s <Wierd Filename>\nFehler - keine Datei angegeben:\n\t%s <Wierd Dateiname>\n",
+	#endif
+		argv[0],argv[0]);
 	 exit (1);
 	}
  infile = fopen (argv[1], "r");
  if (infile == NULL)		/* Bad input filename given */
 	{
+	#ifdef _AMIGA
+	 fprintf (stderr, "Wierd interpreter, Amiga version\n\nError - Invalid Wierd filename: \"%s\".\nFehler:  Ung¸ltiger Wierd-Dateiname: \"%s\".\n",
+	#else
 	 fprintf (stderr, "Wierd interpreter, Atari version\n\nError - Invalid Wierd filename: \"%s\".\nFehler:  UngÅltiger Wierd-Dateiname: \"%s\".\n",
+	#endif
 		argv[1],argv[1]);
 	 exit (2);
 	}
