diff --git a/libvorbis-1.3.7/lib/block.c b/libvorbis-1.3.7/lib/block.c
index 6a50da0..3b39301 100644
--- a/libvorbis-1.3.7/lib/block.c
+++ b/libvorbis-1.3.7/lib/block.c
@@ -417,7 +417,11 @@ static void _preextrapolate_helper(vorbis_dsp_state *v){
   int i;
   int order=16;
   float *lpc=alloca(order*sizeof(*lpc));
+#ifdef __amigaos4__
+  float *work=_ogg_malloc(v->pcm_current*sizeof(*work));
+#else
   float *work=alloca(v->pcm_current*sizeof(*work));
+#endif
   long j;
   v->preextrapolate=1;
 
@@ -452,6 +456,10 @@ static void _preextrapolate_helper(vorbis_dsp_state *v){
 
     }
   }
+
+#ifdef __amigaos4__
+  _ogg_free(work);
+#endif
 }
 
 
